### PR TITLE
Feat/ica proto

### PR DESCRIPTION
**Pull request by seantking**
## Description
Moving the proto files from the `interchain-accounts` repo to `the ibc-go` repo. I also updated any references of IBCAccount -> InterchainAccount.

I also needed to add some additional cosmos `third_party_proto` files to accommodate `account.proto`. 

closes: https://github.com/cosmos/ibc-go/issues/283

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: proto/ibc/applications/interchain_accounts/v1/account.proto [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+// An InterchainAccount is defined as a BaseAccount & the address of the account owner on the controller chain
+message InterchainAccount {
+  option (gogoproto.goproto_getters)         = false;
+  option (gogoproto.goproto_stringer)        = false;
+  option (cosmos_proto.implements_interface) = "InterchainAccountI";
+
+  cosmos.auth.v1beta1.BaseAccount base_account = 1
+      [(gogoproto.embed) = true, (gogoproto.moretags) = "yaml:\"base_account\""];
+  string account_owner = 2 [(gogoproto.moretags) = "yaml:\"port_id\""];
```
Comment (colin-axner) [inline proto/ibc/applications/interchain_accounts/v1/account.proto:18]: fix yaml

File: proto/ibc/applications/interchain_accounts/v1/query.proto [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+// Query request for an interchain account address
+message QueryInterchainAccountRequest {
+  option (gogoproto.equal)           = false;
+  option (gogoproto.goproto_getters) = false;
+
+  // Address is the address to query.
+  string address       = 1;
+  string connection_id = 2 [(gogoproto.moretags) = "yaml:\"connection_id\""];
+}
+
```
Comment (colin-axner) [inline proto/ibc/applications/interchain_accounts/v1/query.proto:25]: we can remove yaml in query.proto files (I'm unsure why, but it is what the SDK does)

File: proto/ibc/applications/interchain_accounts/v1/query.proto [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import "gogoproto/gogo.proto";
+import "google/api/annotations.proto";
+import "ibc/applications/interchain_accounts/v1/account.proto";
+
+option go_package = "github.com/cosmos/ibc-go/modules/apps/27-interchain-accounts/types";
+
+// Query defines the gRPC querier service.
+service Query {
+  // Query to get the address of an interchain account
+  rpc InterchainAccount(QueryInterchainAccountRequest) returns (QueryInterchainAccountResponse) {}
```
Comment (colin-axner) [inline proto/ibc/applications/interchain_accounts/v1/query.proto:13]: Maybe rename to `InterchainAccountAddress`?

File: proto/ibc/applications/interchain_accounts/v1/query.proto [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+// Query request for an interchain account address
+message QueryInterchainAccountRequest {
+  option (gogoproto.equal)           = false;
+  option (gogoproto.goproto_getters) = false;
+
+  // Address is the address to query.
+  string address       = 1;
```
Comment (colin-axner) [inline proto/ibc/applications/interchain_accounts/v1/query.proto:22]: What address is being specified in the request? What address is being returned in the response?

File: proto/ibc/applications/interchain_accounts/v1/types.proto [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  repeated google.protobuf.Any messages = 1;
+}
+
+// The different types of interchain account transactions
+// EXECUTE_TX is used when sending a TX from the controller side to the host side. The host side will execute the tx on
+// behalf of the interchain account.
+enum Type {
+  option (gogoproto.goproto_enum_prefix) = false;
+  // Execute message type
+  TYPE_EXECUTE_TX_UNSPECIFIED = 0 [(gogoproto.enumvalue_customname) = "EXECUTE_TX"];
```
Comment (colin-axner) [inline proto/ibc/applications/interchain_accounts/v1/types.proto:24]: Will there ever be more than just execute tx?

File: proto/ibc/applications/interchain_accounts/v1/types.proto [hunk 6]
```diff
@@ -0,0 +1,31 @@
+syntax = "proto3";
+package ibc.applications.interchain_accounts.v1;
+
+import "google/protobuf/any.proto";
+import "gogoproto/gogo.proto";
+option go_package = "github.com/cosmos/ibc-go/modules/apps/27-interchain-accounts/types";
+
+// Raw tx body
+message IBCTxRaw {
```
Comment (colin-axner) [inline proto/ibc/applications/interchain_accounts/v1/types.proto:9]: What is this used for?

**Review by colin-axner — APPROVED** [review]
The migration part LGTM, but I thought this was a good time to review the proto files as well. We can address those in a followup pr if you'd like. Might be easier

File: proto/ibc/applications/interchain_accounts/v1/types.proto (hunk 5 above)
Comment (seantking) [reply proto/ibc/applications/interchain_accounts/v1/types.proto:24]: Potentially no, but I think it's worth leaving in for extensibility.

**Comment by seantking** [comment]
> The migration part LGTM, but I thought this was a good time to review the proto files as well. We can address those in a followup pr if you'd like. Might be easier

Yep, defo good for discussion. I agree there is much room for improvement. 

I would like to get this merged in sooner rather than later though as I need this in order to hook up the interchain-accounts repo to the `interchain-accounts` branch on the `ibc-go` repo. That will give us the dev environment on the `interchain-accounts` repo & also be a working example for any alpha testers. 

However, I'm going to point it to this branch for now, just to test out everything is working. Doing this atm.

File: proto/ibc/applications/interchain_accounts/v1/query.proto (hunk 3 above)
Comment (seantking) [reply proto/ibc/applications/interchain_accounts/v1/query.proto:13]: Agree.

File: proto/ibc/applications/interchain_accounts/v1/account.proto (hunk 1 above)
Comment (seantking) [reply proto/ibc/applications/interchain_accounts/v1/account.proto:18]: good catch.

File: proto/ibc/applications/interchain_accounts/v1/query.proto (hunk 4 above)
Comment (seantking) [reply proto/ibc/applications/interchain_accounts/v1/query.proto:22]: Good point. The comment is useless here.

File: proto/ibc/applications/interchain_accounts/v1/types.proto (hunk 6 above)
Comment (seantking) [reply proto/ibc/applications/interchain_accounts/v1/types.proto:9]: We are using this to unmarshal data (SDK messages) sent in the IBC transaction -> https://github.com/cosmos/ibc-go/blob/interchain-accounts/modules/apps/27-interchain-accounts/keeper/relay.go#L97

See here also -> https://github.com/cosmos/ibc-go/blob/interchain-accounts/modules/apps/27-interchain-accounts/keeper/relay.go#L210

File: proto/ibc/applications/interchain_accounts/v1/types.proto (hunk 5 above)
Comment (colin-axner) [reply proto/ibc/applications/interchain_accounts/v1/types.proto:24]: Yea I just figure it isn't backwards compatible anyways. If someone uses an enum with value 1, the current implementation couldn't do anything about it and would return an error

File: proto/ibc/applications/interchain_accounts/v1/types.proto (hunk 6 above)
Comment (colin-axner) [reply proto/ibc/applications/interchain_accounts/v1/types.proto:9]: It just seems like an unnecessary wrapper. Don't we just need one type with an array of Any's. Then we marshal/unmarshal that and pass the marshaled version as `[]byte` in the packet?

**Comment by colin-axner** [comment]
Feel free to merge whenever you think is good @seantking :+1:

File: proto/ibc/applications/interchain_accounts/v1/account.proto (hunk 1 above)
Comment (seantking) [reply proto/ibc/applications/interchain_accounts/v1/account.proto]: Updated this.

File: proto/ibc/applications/interchain_accounts/v1/query.proto (hunk 2 above)
Comment (seantking) [reply proto/ibc/applications/interchain_accounts/v1/query.proto:25]: Updated.

File: proto/ibc/applications/interchain_accounts/v1/query.proto (hunk 3 above)
Comment (seantking) [reply proto/ibc/applications/interchain_accounts/v1/query.proto]: Updated

File: proto/ibc/applications/interchain_accounts/v1/query.proto (hunk 4 above)
Comment (seantking) [reply proto/ibc/applications/interchain_accounts/v1/query.proto]: I updated the comments + param names. Is it more clear?

File: proto/ibc/applications/interchain_accounts/v1/types.proto (hunk 5 above)
Comment (seantking) [reply proto/ibc/applications/interchain_accounts/v1/types.proto:24]: Yeah, it's being passed in the packet. The packet is made up of type + data bytes. I'll think about it though, good points.

File: proto/ibc/applications/interchain_accounts/v1/types.proto (hunk 6 above)
Comment (seantking) [reply proto/ibc/applications/interchain_accounts/v1/types.proto:9]: You are correct. I think this is left over from when there was a concept of having different types of deserialization for different types of chains. Seeing as we are only focusing on SDK chains for now we can probably remove it. I'll do it in a follow-up PR. 

I'll make a few issues out of the discussions here for leftover work.

File: proto/ibc/applications/interchain_accounts/v1/genesis.proto [hunk 7]
```diff
@@ -0,0 +1,11 @@
+syntax = "proto3";
+package ibc.applications.interchain_accounts.v1;
+
+option go_package = "github.com/cosmos/ibc-go/modules/apps/27-interchain-accounts/types";
+
+import "gogoproto/gogo.proto";
+
+// GenesisState defines the interchain_account genesis state
+message GenesisState {
+  string port_id = 1 [(gogoproto.moretags) = "yaml:\"port_id\""];
```
Comment (seantking) [inline proto/ibc/applications/interchain_accounts/v1/genesis.proto:10]: @colin-axner I am not entirely sure what we want to put in our genesis state. :thinking:

Comment (colin-axner) [reply proto/ibc/applications/interchain_accounts/v1/genesis.proto:10]: Any state that needs to be persisted between dump state/restart upgrades

**Pull request merged**